Libraries/Opc.Ua.Client/NodeCache.cs

```diff
@@ -612,33 +660,43 @@ public Node FetchNode(ExpandedNodeId nodeId)
                 // fetch references from server.
                 ReferenceDescriptionCollection references = m_session.FetchReferences(localId);
 
-                foreach (ReferenceDescription reference in references)
+                try
                 {
-                    // create a placeholder for the node if it does not already exist.
-                    if (!m_nodes.Exists(reference.NodeId))
+                    m_cacheLock.EnterWriteLock();
```

Comment:
Start as read (99% case) and upgrade lock to write when adding? Is that possible?

Reply:
Yes, I checked the attach code a bit closer, it is sufficient to write protect the Attach function, as it is replacing duplicates. Then upgrade is a good choice. However I noticed that access to m_typetree also needs a read lock, because it is updated by attach